src/types.ts

```diff
@@ -11,21 +16,40 @@ interface Refactoring {
   };
 }
 
+interface ActionProvider {
+  message: string;
+  isPreferred?: boolean;
+  createVisitor: (
+    selection: Selection,
+    onMatch: (path: NodePath<any>) => void,
+    refactoring: RefactoringWithActionProvider
+  ) => Visitor;
+  updateMessage?: (path: NodePath<any>) => void;
```

Comment:
🍰 Suggestion: make this return a new message string, and update the message with this value from the generic implementation.

Otherwise it's called `updateMessage()` but really it could do anything. I'd prefer to enforce the behavior.

Finally, what do you think about merging this with the `message` one? We could have a `message` attributes that could either be a `string` or a function returning a `string`.

Reply:
I liked your suggestion to merge `updateMessage()` and `message` but they are accessed at different places (`updateMessage()` is called during the AST traversal, `message` is read afterwards). I'll keep the `updateMessage()` function for now but make it return `string`.